group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -34,48 +33,68 @@
- *   still owned by other members in the group.
+ * - UNACKNOWLEDGED_ASSIGNMENT:
+ *   The member has received a new assignment from the group coordinator but
+ *   he has not acknowledged it yet. The member is removed from the group if
+ *   he does not acknowledge it within the rebalance timeout.
  *
- * The state machine has three states:
- * - REVOKING:
- *   This state means that the member must revoke partitions before it can transition to the next epoch
- *   and thus start receiving new partitions. This is to guarantee that offsets of revoked partitions
- *   are committed with the current epoch. The member transitions to the next state only when it has
- *   acknowledged the revocation.
+ *   Valid Transitions:
+ *   - UNACKNOWLEDGED_ASSIGNMENT -> STABLE
+ *     When the assignment is acknowledged, the member transitions to the STABLE
+ *     state if it is fully reconciled.
  *
- * - ASSIGNING:
- *   This state means that the member waits on partitions which are still owned by other members in the
- *   group. It remains in this state until they are all freed up.
+ *   - UNACKNOWLEDGED_ASSIGNMENT -> UNACKNOWLEDGED_ASSIGNMENT
+ *     When the assignment is acknowledged, the member remains in the
+ *     UNACKNOWLEDGED_ASSIGNMENT state if a new assignment is computed.
  *
- * - STABLE:
- *   This state means that the member has received all its assigned partitions.
+ *     If the next assignment contains partitions to be revoked, the member stays
+ *     in his current epoch. Otherwise, he transitions to the target epoch.
+ *
+ *   - UNACKNOWLEDGED_ASSIGNMENT -> UNRELEASED_PARTITIONS
+ *     When the assignment is acknowledged, the member transitions to the
+ *     UNRELEASED_PARTITIONS if newly assigned partitions are not available yet.
+ *
+ * - UNRELEASED_PARTITIONS:
+ *   The member's reconciliation cannot progress because newly assigned partitions
+ *   are still owned by other members in the group. They are not released yet.
  *
- * The reconciliation process is started or re-started whenever a new target assignment is installed;
- * the epoch of the new target assignment is different from the next epoch of the member. In this transient
- * state, the assigned partitions, the partitions pending revocation and the partitions pending assignment
- * are updated. If the partitions pending revocation is not empty, the state machine transitions to
- * REVOKING; if partitions pending assignment is not empty, it transitions to ASSIGNING; otherwise it
- * transitions to STABLE.
+ *   Valid Transitions:
+ *   - UNRELEASED_PARTITIONS -> STABLE
+ *     The member may transition to the STABLE state if the partitions that he
+ *     was awaiting on are no longer in the desired target assignment.
+ *
+ *   - UNRELEASED_PARTITIONS -> UNACKNOWLEDGED_ASSIGNMENT
+ *     When at least one partition become available, a new assignment is computed
+ *     for the member and he transitions to the UNACKNOWLEDGED_ASSIGNMENT state.
```

Comment:
are there any implications to computing a new assignment for every acknowledged partitions?

Reply:
Why are we computing a new assignment when the partition becomes available? Wouldn't we just assign it since it was already part of our desired assignment and it became available?

Reply:
Do correct me if I'm wrong David, so we send partitions to be assigned and partitions to be revoked to the client. The ack is received when there's more partitions available because they were revoked and now when we "compute" it doesn't necessarily mean that a whole new assignment was computed by the assignor, it could just be the newly released partitions and pending partitions to revoke.